oracle.go

```diff
@@ -225,30 +226,32 @@ func (o *DBOracle) cleanup() {
 // CompletedTokens looks at the list of tokens and finds what referenced txns
 // are completed, and then returns the set of tokens that are completed.
 func (o *DBOracle) CompletedTokens(tokens []string) (map[string]bool, error) {
-	objectIds := make([]bson.ObjectId, 0, len(tokens))
+	objectIds := make(map[bson.ObjectId]bool, len(tokens))
 
 	// The nonce is generated during preparing, and if 2 flushers race,
 	// only one nonce makes it into the final transaction. However, other
 	// nonces can also be considered 'completed'. (afaict, they are ignored,
 	// thus won't be applied and can be considered completed.)
 	for _, token := range tokens {
 		objId := txnTokenToId(token)
-		objectIds = append(objectIds, objId)
+		objectIds[objId] = true
 	}
-	query := o.working.Find(bson.M{"_id": bson.M{"$in": objectIds}})
-	query = query.Select(bson.M{"_id": 1})
-	iter := query.Iter()
+
+	iter := o.working.Find(nil).Select(bson.M{"_id": 1}).Iter()
```

Comment:
Sorry about the delay in responding to this.
I'm a little concerned that o.working can be Millions of rows (in our worst case times, we've seen 50M records to be cleaned up).
And while a given doc can be larger than we'd like (300k, IIRC), in the common case the list of tokens is going to be quite small, and this has to iterate 50M records for every document with <10 tokens.

I would be ok with having an if/else here, something around:
if len(objectIds) > 1000 {
  // mongo is known to be slow when handling large $in queries, so we switch format
  iter = 
} else {
  iter = o.working....
}